File: bindings/php/commands.php
<?php

include "uhppote/uhppote.php";

define('CONTROLLER', 405419896);
define('DOOR', 3);
define('MODE', 2);
define('DELAY', 10);
define('CARD', 10058400);
define('CARD_INDEX', 3);
define('EVENT_INDEX', 37);
define('TIME_PROFILE_ID', 29);
define('LISTENER_INTERVAL', 15);

define('ADDRESS', '192.168.1.100');
define('NETMASK', '255.255.255.0');
define('GATEWAY', '192.168.1.1');
define('LISTENER', '192.168.1.100:60001');

function execute($cmd, $fn, $options, $args)
{
    $uhppote = new \uhppote\UHPPOTE(
        $options['bind'],
        $options['broadcast'],
        $options['listen'],
        $options['timeout'],
        $options['debug']
    );

    try {
        pprint($cmd, $fn($uhppote, $args));
    } catch (Exception $e) {
        echo "\n   *** ERROR:  ",  $e->getMessage(), "\n\n";
    }
}

function pprint($cmd, $result)
{
    if ($cmd != '') {
        print("   $cmd\n");
    }

    if (is_array($result)) {
        foreach ($result as $response) {
            pprint('', $response);
        }
    } else {
        $width = 0;
        foreach ($result as $key => $value) {
            if (strlen($key) > $width) {
                $width = strlen($key);
            }
        }

        $format = sprintf("      %%-%ds  %%s\n", $width);

        foreach ($result as $key => $value) {
            if (is_bool($value)) {
                printf($format, $key, $value ? 'Y' : 'N');
            } else {
                printf($format, $key, $value);
            }
        }
    }

    print("\n");
}

function get_all_controllers($u, $args)
{
    return uhppote\get_all_controllers($u);
}

function get_controller($u, $args)
{
    $controller = resolve(CONTROLLER);

    return uhppote\get_controller($u, $controller);
}

function set_ipv4($u, $args)
{
    $controller = resolve(CONTROLLER);
    $address = ADDRESS;
    $netmask = NETMASK;
    $gateway = GATEWAY;

    uhppote\set_ipv4($u, $controller, $address, $netmask, $gateway);

    return (object) array('set' => 'ok');
}

function get_time($u, $args)
{
    $controller = resolve(CONTROLLER);

    return uhppote\get_time($u, $controller);
}

function set_time($u, $args)
{
    $controller = resolve(CONTROLLER);
    $now = new DateTimeImmutable('now', new DateTimeZone('PDT'));

    return uhppote\set_time($u, $controller, $now);
}

function get_listener($u, $args)
{
    $controller = resolve(CONTROLLER);
    return uhppote\get_listener($u, $controller);
}

function set_listener($u, $args)
{
    $controller = resolve(CONTROLLER);
    list($address, $port) = explode(':', LISTENER);
    $interval = LISTENER_INTERVAL;

    return uhppote\set_listener($u, $controller, $address, (int)$port, (int) $interval);
}

function get_door($u, $args)
{
    $controller = resolve(CONTROLLER);
    $door = DOOR;

    return uhppote\get_door($u, $controller, $door);
}

function set_door_control($u, $args)
{
    $controller = resolve(CONTROLLER);
    $door = DOOR;
    $mode = MODE;
    $delay = DELAY;

    return uhppote\set_door_control($u, $controller, $door, $mode, $delay);
}

function get_status($u, $args)
{
    $controller = resolve(CONTROLLER);

    return uhppote\get_status($u, $controller);
}

function open_door($u, $args)
{
    $controller = resolve(CONTROLLER);
    $door = DOOR;

    return uhppote\open_door($u, $controller, $door);
}

function get_cards($u, $args)
{
    $controller = resolve(CONTROLLER);

    return uhppote\get_cards($u, $controller);
}

function get_card($u, $args)
{
    $controller = resolve(CONTROLLER);
    $card = CARD;

    return uhppote\get_card($u, $controller, $card);
}

function get_card_by_index($u, $args)
{
    $controller = resolve(CONTROLLER);
    $index = CARD_INDEX;

    return uhppote\get_card_by_index($u, $controller, $index);
}

function put_card($u, $args)
{
    $controller = resolve(CONTROLLER);
    $card = CARD;
    $start = new DateTimeImmutable('2023-01 -01', new DateTimeZone('PDT'));
    $end = new DateTimeImmutable('2023-12-31', new DateTimeZone('PDT'));
    $door1 = 0;
    $door2 = 1;
    $door3 = 29;
    $door4 = 0;
    $pin = 7531;

    return uhppote\put_card($u, $controller, $card, $start, $end, $door1, $door2, $door3, $door4, $pin);
}

function delete_card($u, $args)
{
    $controller = resolve(CONTROLLER);
    $card = CARD;

    return uhppote\delete_card($u, $controller, $card);
}

function delete_all_cards($u, $args)
{
    $controller = resolve(CONTROLLER);

    return uhppote\delete_all_cards($u, $controller);
}

function get_event($u, $args)
{
    $controller = resolve(CONTROLLER);
    $index = EVENT_INDEX;

    return uhppote\get_event($u, $controller, $index);
}

function get_event_index($u, $args)
{
    $controller = resolve(CONTROLLER);

    return uhppote\get_event_index($u, $controller);
}

function set_event_index($u, $args)
{
    $controller = resolve(CONTROLLER);
    $index = EVENT_INDEX;

    return uhppote\set_event_index($u, $controller, $index);
}


function record_special_events($u, $args)
{
    $controller = resolve(CONTROLLER);
    $enabled = true;

    return uhppote\record_special_events($u, $controller, $enabled);
}

function get_time_profile($u, $args)
{
    $controller = resolve(CONTROLLER);
    $profile_id = TIME_PROFILE_ID;

    return uhppote\get_time_profile($u, $controller, $profile_id);
}

function set_time_profile($u, $args)
{
    $controller = resolve(CONTROLLER);
    $profile_id = TIME_PROFILE_ID;
    $start = new DateTimeImmutable('2023-01 -01', new DateTimeZone('PDT'));
    $end = new DateTimeImmutable('2023-12-31', new DateTimeZone('PDT'));
    $monday = true;
    $tuesday = true;
    $wednesday = false;
    $thursday = true;
    $friday = false;
    $saturday = false;
    $sunday = true;
    $segment1_start = new DateTimeImmutable('08:30', new DateTimeZone('PDT'));
    $segment1_end = new DateTimeImmutable('11:45', new DateTimeZone('PDT'));
    $segment2_start = new DateTimeImmutable('13:15', new DateTimeZone('PDT'));
    $segment2_end = new DateTimeImmutable('16:30', new DateTimeZone('PDT'));
    $segment3_start = new DateTimeImmutable('19:30', new DateTimeZone('PDT'));
    $segment3_end = new DateTimeImmutable('20:55', new DateTimeZone('PDT'));
    $linked_profile_id = 30;


    return uhppote\set_time_profile(
        $u,
        $controller,
        $profile_id,
        $start,
        $end,
        $monday,
        $tuesday,
        $wednesday,
        $thursday,
        $friday,
        $saturday,
        $sunday,
        $segment1_start,
        $segment1_end,
        $segment2_start,
        $segment2_end,
        $segment3_start,
        $segment3_end,
        $linked_profile_id
    );
}

function delete_all_time_profiles($u, $args)
{
    $controller = resolve(CONTROLLER);

    return uhppote\delete_all_time_profiles($u, $controller);
}

function add_task($u, $args)
{
    $controller = resolve(CONTROLLER);
    $start_date = new DateTimeImmutable('2023-01 -01', new DateTimeZone('PDT'));
    $end_date = new DateTimeImmutable('2023-12-31', new DateTimeZone('PDT'));
    $monday = true;
    $tuesday = true;
    $wednesday = false;
    $thursday = true;
    $friday = false;
    $saturday = false;
    $sunday = true;
    $start_time = new DateTimeImmutable('08:30', new DateTimeZone('PDT'));
    $door = DOOR;
    $task_type = 2;
    $more_cards = 0;

    return uhppote\add_task(
        $u,
        $controller,
        $start_date,
        $end_date,
        $monday,
        $tuesday,
        $wednesday,
        $thursday,
        $friday,
        $saturday,
        $sunday,
        $start_time,
        $door,
        $task_type,
        $more_cards
    );
}

function refresh_tasklist($u, $args)
{
    $controller = resolve(CONTROLLER);

    return uhppote\refresh_tasklist($u, $controller);
}

function clear_tasklist($u, $args)
{
    $controller = resolve(CONTROLLER);

    return uhppote\clear_tasklist($u, $controller);
}

function set_pc_control($u, $args)
{
    $controller = resolve(CONTROLLER);
    $enabled = true;

    return uhppote\set_pc_control($u, $controller, $enabled);
}

function set_interlock($u, $args)
{
    $controller = resolve(CONTROLLER);
    $interlock = 3;

    return uhppote\set_interlock($u, $controller, $interlock);
}

function activate_keypads($u, $args)
{
    $controller = resolve(CONTROLLER);
    $reader1 = true;
    $reader2 = true;
    $reader3 = false;
    $reader4 = true;

    return uhppote\activate_keypads($u, $controller, $reader1, $reader2, $reader3, $reader4);
}

function set_door_passcodes($u, $args)
{
    $controller = resolve(CONTROLLER);
    $door = DOOR;
    $passcode1 = 12345;
    $passcode2 = 0;
    $passcode3 = 999999;
    $passcode4 = 54321;

    return uhppote\set_door_passcodes($u, $controller, $door, $passcode1, $passcode2, $passcode3, $passcode4);
}

function get_antipassback($u, $args)
{
    $controller = resolve(CONTROLLER);

    return uhppote\get_antipassback($u, $controller);
}

function set_antipassback($u, $args)
{
    $controller = resolve(CONTROLLER);
    $antipassback = 2;

    return uhppote\set_antipassback($u, $controller, $antipassback);
}

function restore_default_parameters($u, $args)
{
    $controller = resolve(CONTROLLER);

    return uhppote\restore_default_parameters($u, $controller);
}

function listen($u, $args)
{
    uhppote\listen($u, function ($event) {
        pprint('event', $event);
    });
}

function resolve($controller)
{ 
    $list = controllers();

    if (isset($list[$controller])) {
        return $list[$controller];
    }

    return new \uhppote\Controller($controller, '', 'udp');
}

function commands()
{
    return  [
        'get-all-controllers' => 'get_all_controllers',
        'get-controller' => 'get_controller',
        'set-IPv4' => 'set_ipv4',
        'get-time' => 'get_time',
        'set-time' => 'set_time',
        'get-listener' => 'get_listener',
        'set-listener' => 'set_listener',
        'get-door' => 'get_door',
        'set-door-control' => 'set_door_control',
        'get-status' => 'get_status',
        'open-door' => 'open_door',
        'get-cards' => 'get_cards',
        'get-card' => 'get_card',
        'get-card-by-index' => 'get_card_by_index',
        'put-card' => 'put_card',
        'delete-card' => 'delete_card',
        'delete-all-cards' => 'delete_all_cards',
        'get-event' => 'get_event',
        'get-event-index' => 'get_event_index',
        'set-event-index' => 'set_event_index',
        'record-special-events' => 'record_special_events',
        'get-time-profile' => 'get_time_profile',
        'set-time-profile' => 'set_time_profile',
        'delete-all-time-profiles' => 'delete_all_time_profiles',
        'add-task' => 'add_task',
        'refresh-tasklist' => 'refresh_tasklist',
        'clear-tasklist' => 'clear_tasklist',
        'set-pc-control' => 'set_pc_control',
        'set-interlock' => 'set_interlock',
        'activate-keypads' => 'activate_keypads',
        'set-door-passcodes' => 'set_door_passcodes',
        'get-antipassback' => 'get_antipassback',
        'set-antipassback' => 'set_antipassback',
        'restore-default-parameters' => 'restore_default_parameters',
        'listen' => 'listen'
    ];
}

function controllers()
{
    return array(
        405419896 => new \uhppote\Controller(405419896, '192.168.1.100:60000', 'tcp'),
        303986753 => new \uhppote\Controller(303986753, '192.168.1.100:60000', 'udp'),
    );
}

File: bindings/python/commands.py
import ipaddress
import datetime

import uhppote

from pprint import pprint

CONTROLLER = 405419896
DOOR = 3
MODE = 2
DELAY = 10
CARD = 8165538
CARD_INDEX = 3
EVENT_INDEX = 37
TIME_PROFILE_ID = 29
LISTENER_INTERVAL = 15

ADDRESS = ipaddress.IPv4Address('192.168.1.100')
NETMASK = ipaddress.IPv4Address('255.255.255.0')
GATEWAY = ipaddress.IPv4Address('192.168.1.1')
LISTENER = (ipaddress.IPv4Address('192.168.1.100'), 60001)

CONTROLLERS = {
     405419896: uhppote.Controller(405419896,'192.168.1.100:60000','tcp'),
}


def commands():
    return {
        'get-all-controllers': get_all_controllers,
        'get-controller': get_controller,
        'set-IPv4': set_IPv4,
        'get-time': get_time,
        'set-time': set_time,
        'get-listener': get_listener,
        'set-listener': set_listener,
        'get-door': get_door,
        'set-door': set_door,
        'get-status': get_status,
        'open-door': open_door,
        'get-cards': get_cards,
        'get-card': get_card,
        'get-card-at-index': get_card_at_index,
        'put-card': put_card,
        'delete-card': delete_card,
        'delete-all-cards': delete_all_cards,
        'get-event': get_event,
        'get-event-index': get_event_index,
        'set-event-index': set_event_index,
        'record-special-events': record_special_events,
        'get-time-profile': get_time_profile,
        'set-time-profile': set_time_profile,
        'clear-time-profiles': clear_time_profiles,
        'add-task': add_task,
        'refresh-tasklist': refresh_tasklist,
        'clear-tasklist': clear_tasklist,
        'set-pc-control': set_pc_control,
        'set-interlock': set_interlock,
        'activate-keypads': activate_keypads,
        'set-door-passcodes': set_door_passcodes,
        'get-antipassback': get_antipassback,
        'set-antipassback': set_antipassback,
        'restore-default-parameters': restore_default_parameters,
        'listen': listen,
    }


def exec(f, bind, broadcast, listen, debug):
    u = uhppote.Uhppote(bind, broadcast, listen, debug)
    response = f(u)

    if response != None:
        if type(response).__name__ == 'list':
            for v in response:
                pprint(v.__dict__, indent=2, width=1, sort_dicts=False)
        elif type(response).__name__ == 'bool':
            pprint(response, indent=2, width=1, sort_dicts=False)
        else:
            pprint(response.__dict__, indent=2, width=1, sort_dicts=False)


def get_all_controllers(u):
    return u.get_all_controllers()


def get_controller(u):
    controller = resolve(CONTROLLER)

    return u.get_controller(controller)


def set_IPv4(u):
    controller = resolve(CONTROLLER)
    address = ADDRESS
    netmask = NETMASK
    gateway = GATEWAY

    return u.set_ipv4(controller, address, netmask, gateway)


def get_time(u):
    controller = resolve(CONTROLLER)

    return u.get_time(controller)


def set_time(u):
    controller = resolve(CONTROLLER)
    now = datetime.datetime.now()

    return u.set_time(controller, now)


def get_listener(u):
    controller = resolve(CONTROLLER)

    return u.get_listener(controller)


def set_listener(u):
    controller = resolve(CONTROLLER)
    (address, port) = LISTENER
    interval = LISTENER_INTERVAL

    return u.set_listener(controller, address, port, interval)


def get_door(u):
    controller = resolve(CONTROLLER)
    door = DOOR

    return u.get_door(controller, door)


def set_door(u):
    controller = resolve(CONTROLLER)
    door = DOOR
    mode = MODE
    delay = DELAY

    return u.set_door(controller, door, mode, delay)


def get_status(u):
    controller = resolve(CONTROLLER)

    return u.get_status(controller)


def open_door(u):
    controller = resolve(CONTROLLER)
    door = DOOR

    return u.open_door(controller, door)


def get_cards(u):
    controller = resolve(CONTROLLER)

    return u.get_cards(controller)


def get_card(u):
    controller = resolve(CONTROLLER)
    card = CARD

    response = u.get_card(controller, card)
    if response.card == 0:
        raise ValueError(f'card {card} not found')

    return response


def get_card_at_index(u):
    controller = resolve(CONTROLLER)
    index = CARD_INDEX

    response = u.get_card_at_index(controller, index)
    if response.card == 0:
        raise ValueError(f'card @ index {index} not found')
    elif response.card == 0xffffffff:
        raise ValueError(f'card @ index {index} deleted')

    return response


def put_card(u):
    controller = resolve(CONTROLLER)
    card = CARD
    start = datetime.datetime.strptime("2022-01-01", '%Y-%m-%d').date()
    end = datetime.datetime.strptime("2022-12-31", '%Y-%m-%d').date()

    return u.put_card(controller, card, start, end, 0, 1, 29, 0, 7531)


def delete_card(u):
    controller = resolve(CONTROLLER)
    card = CARD

    return u.delete_card(controller, card)


def delete_all_cards(u):
    controller = resolve(CONTROLLER)

    return u.delete_all_cards(controller)


def get_event(u):
    controller = resolve(CONTROLLER)
    index = EVENT_INDEX

    response = u.get_event(controller, index)
    if response.event_type == 0xff:
        raise ValueError(f'event @ index {index} overwritten')
    elif response.index == 0:
        raise ValueError(f'event @ index {index} not found')

    return response


def get_event_index(u):
    controller = resolve(CONTROLLER)

    return u.get_event_index(controller)


def set_event_index(u):
    controller = resolve(CONTROLLER)
    index = EVENT_INDEX

    return u.set_event_index(controller, index)


def record_special_events(u):
    controller = resolve(CONTROLLER)
    enabled = True

    return u.record_special_events(controller, enabled)


def get_time_profile(u):
    controller = resolve(CONTROLLER)
    profile_id = TIME_PROFILE_ID

    response = u.get_time_profile(controller, profile_id)
    if response.profile_id == 0:
        raise ValueError(f'time profile {profile_id} not defined')

    return response


def set_time_profile(u):
    controller = resolve(CONTROLLER)
    profile_id = TIME_PROFILE_ID
    start = datetime.datetime.strptime("2022-01-01", '%Y-%m-%d').date()
    end = datetime.datetime.strptime("2022-12-31", '%Y-%m-%d').date()
    monday = True
    tuesday = False
    wednesday = True
    thursday = True
    friday = False
    saturday = False
    sunday = True
    segment1start = datetime.datetime.strptime("08:15", '%H:%M').time()
    segment1end = datetime.datetime.strptime("11:45", '%H:%M').time()
    segment2start = datetime.datetime.strptime("12:45", '%H:%M').time()
    segment2end = datetime.datetime.strptime("17:15", '%H:%M').time()
    segment3start = datetime.datetime.strptime("19:30", '%H:%M').time()
    segment3end = datetime.datetime.strptime("22:00", '%H:%M').time()
    linked_profile_ID = 23

    return u.set_time_profile(
        controller, 
        profile_id,
        start, end,
        monday, tuesday, wednesday, thursday, friday, saturday, sunday,
        segment1start, segment1end,
        segment2start, segment2end,
        segment3start, segment3end,
        linked_profile_ID)


def clear_time_profiles(u):
    controller = resolve(CONTROLLER)

    return u.clear_time_profiles(controller)


def add_task(u):
    controller = resolve(CONTROLLER)
    start_date = datetime.datetime.strptime("2022-01-01", '%Y-%m-%d').date()
    end_date = datetime.datetime.strptime("2022-12-31", '%Y-%m-%d').date()
    monday = True
    tuesday = False
    wednesday = True
    thursday = True
    friday = False
    saturday = False
    sunday = True
    start_time = datetime.datetime.strptime("08:15", '%H:%M').time()
    door = DOOR
    task_type = 2
    more_cards = 0

    return u.add_task(
        controller,
        task_type,
        start_date, end_date,
        monday, tuesday, wednesday, thursday, friday, saturday, sunday,
        start_time,
        door,
        more_cards)


def refresh_tasklist(u):
    controller = resolve(CONTROLLER)

    return u.refresh_task_list(controller)


def clear_tasklist(u):
    controller = resolve(CONTROLLER)

    return u.clear_task_list(controller)


def set_pc_control(u):
    controller = resolve(CONTROLLER)
    enabled = True

    return u.set_pc_control(controller, enabled)


def set_interlock(u):
    controller = resolve(CONTROLLER)
    interlock = 3

    return u.set_interlock(controller, interlock)


def activate_keypads(u):
    controller = resolve(CONTROLLER)
    reader1 = True
    reader2 = True
    reader3 = False
    reader4 = True

    return u.activate_keypads(controller, reader1, reader2, reader3, reader4)


def set_door_passcodes(u):
    controller = resolve(CONTROLLER)
    door = DOOR
    passcode1 = 12345
    passcode2 = 0
    passcode3 = 999999
    passcode4 = 54321

    return u.set_door_passcodes(controller, door, passcode1, passcode2, passcode3, passcode4)


def get_antipassback(u):
    controller = resolve(CONTROLLER)

    return u.get_antipassback(controller)


def set_antipassback(u):
    controller = resolve(CONTROLLER)
    antipassback = 2

    return u.set_antipassback(controller, antipassback)


def restore_default_parameters(u):
    controller = resolve(CONTROLLER)

    return u.restore_default_parameters(controller)


def listen(u):
    return u.listen(onEvent)


def onEvent(event):
    if event != None:
        pprint(event.__dict__, indent=2, width=1)

def resolve(controller):
    v = CONTROLLERS.get(controller, None)
    if not v is None:
        return v
    else:
        return uhppote.Controller(controller, None, 'udp')
